authentication.py

```diff
@@ -0,0 +1,84 @@
+        self.id = id
+        self.email = email
+        self.token = token
+        self.expires_in = expires_in
+
+    def __str__(self):
+        return 'id: {}, email: {}, token: {}, expires_id: {}'.format(self.id, self.email, self.token, self.expires_in)
+
+
+def _token_info(token):
+    result = urlfetch.fetch(
+        '%s?%s' % (_TOKENINFO_URL, urllib.urlencode({'access_token': token})))
+    if result.status_code != 200:
+        raise endpoints.InternalServerErrorException(message='Token info endpoint returned status {}: {}'.format(result.status_code, result.content))
+    return result
+
+
+def require_user_info(request_state, token_info_fn=_token_info):
+    """Get the user's info from cache or from google if not in cache, throwing unauthorized errors as appropriate
+
+    :param request_state: self.request_state from a class extending protorpc.remote.Service
+    :param token_info_fn: function to get token info, defaults to google http request, override for testing
+    :return: UserInfo instance
+    """
+    auth_header = request_state.headers.get('Authorization')
+    if auth_header is None:
+        raise endpoints.UnauthorizedException(message='Request missing Authorization header.')
+
+    auth_header_parts = auth_header.split()
+    if len(auth_header_parts) != 2 or auth_header_parts[0].lower() != 'bearer':
+        raise endpoints.UnauthorizedException(message='Malformed Authorization header.')
+
+    token = auth_header_parts[1]
+
+    user_info = memcache.get(key=token)
+    if user_info is None:
+        logging.debug("auth token cache miss for token %s", token)
+        user_info = _fetch_user_info(token, token_info_fn)
+        # cache for 10 minutes or until token expires
+        memcache.add(key=token, value=user_info, time=min([user_info.expires_in, 600]))
+
+    return user_info
+
+
+def _fetch_user_info(token, token_info_fn):
+    """Make the external call to get token info and create UserInfo object"""
+    # Get token info from the tokeninfo endpoint.
+    result = token_info_fn(token)
+
+    token_info = json.loads(result.content)
+    logging.debug("token info for %s: %s", token, json.dumps(token_info))
+
+    # Validate email.
+    if 'email' not in token_info:
+        raise endpoints.UnauthorizedException(message='Oauth token doesn\'t include an email address.')
+    if not token_info.get('verified_email'):
+        raise endpoints.UnauthorizedException(message='Oauth token email isn\'t verified.')
+    if 'user_id' not in token_info:
+        raise endpoints.UnauthorizedException(message='Oauth token doesn\'t include user_id.')
+
+    # Validate audience.
+    audience = token_info.get('audience')
+    if any(audience.startswith(prefix) for prefix in _ACCEPTED_AUDIENCE_PREFIXES) or \
+            any(token_info.get('email').endswith(suffix) for suffix in _ACCEPTED_EMAIL_SUFFIXES):
+        return UserInfo(token_info.get('user_id'), token_info.get('email'), token, int(token_info.get('expires_in')))
```

Comment:
Not validating the `expires_in` value, in particular, if it's present.  See other comment when adding the token to the cache